language/move-binary-format/src/compatibility.rs

```diff
@@ -58,27 +64,64 @@ impl CompatibilityConfig {
         }
     }
 
+    pub fn new(
+        check_struct_and_function_linking: bool,
+        check_struct_layout: bool,
+        check_friend_linking: bool,
+    ) -> Self {
+        if check_friend_linking {
+            assert!(
+                check_struct_and_function_linking,
+                "check_friend_linking requires check_struct_and_function_linking"
+            );
+        }
+        Self {
+            check_struct_and_function_linking,
+            check_struct_layout,
+            check_friend_linking,
+        }
+    }
+
     pub fn need_check_compat(&self) -> bool {
         // `check_friend_linking` is part of check_struct_and_function_linking, so ignore it at here.
         self.check_struct_and_function_linking || self.check_struct_layout
     }
+
+    /// Return compatibility assessment for `new_module` relative to old module `old_module`.
+    pub fn check(&self, old_module: &Module, new_module: &Module) -> PartialVMResult<()> {
```

Comment:
Wondering if we can simplify this even further by unifying the `Compatibility` and `CompatibilityInfo` types. That is:

- Make `CompatibilityConfig::new` a function of `Compatibility`
- change the signature of `Compatibility::check` to `check(&self, old_module, new_module)` (no return value)
- Move the `return Err` login in this function inside `Compatibility::check`
- Get rid of the `CompatibilityInfo` struct + functions altogether; in the new design, you define a compatibility policy with `Compatibility::new`, then check it via `Compatibility::check`

Reply:
So, it looks like this:

```
#[derive(Debug, Clone, Copy)]
pub struct Compatibility {
    pub check_struct_and_function_linking: bool,
    pub check_struct_layout: bool,
    pub check_friend_linking: bool,
}

impl Compatibility{

  pub fun new(check_struct_and_function_linking: bool,
        check_struct_layout: bool,
        check_friend_linking: bool,
    ) -> Self;

  pub fn check(&self, old_module: &Module, new_module: &Module) -> PartialVMResult<()>;

}
```

But when to return `CompatibilityInfo`? add a new function to `Compatibility` like:

```
  pub fn check_for_info(old_module: &Module, new_module: &Module) -> CompatibilityInfo;
```

Reply:
I thin we should no longer need `CompatibilityConfig`--all of the old clients can just use `Compatibility` directly.

Reply:
Got it `CompatibilityInfo` is a typo of `CompatibilityConfig`. But sometimes we need to get the `Compatibility` result but no error, such as https://github.com/move-language/move/pull/559/files#diff-8a42ae3ffa92bcdc1150d38c74c7d4feaacadab017ecd3e7934b52d7429f1896R350

Reply:
For that particular call site, I think it's ok to do something like

```
if compat.check(&old_api, &new_api).is_err() {
  // check whether it's a layout or linking issue by constructing a new config and re-checking
  if Compatibility::new(true, false, false).is_err() {
    println!("Layout API for structs of module {} has changed. Need to do a data migration of published structs", module_id)
  }
  if Comptability::new(false, true, true).is_err() {
     println!("Linking API for structs/functions of module {} has changed. Need to redeploy all dependent modules.", module_id)
  }
}
```

Reply:
This is fine, but it feels a bit stupid to check Compatibility twice😂, I'll try it tomorrow.

Reply:
Yep, this is definitely a bit silly :). But it's on an error path, and in service of making the safety-critical usages of the API as simple as possible.

Reply:
Refactor as the comment.